osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/configuration/service/SetConfigurationObjectService.java

```diff
@@ -0,0 +1,125 @@
+        final SetParameter setParameter = new SetParameter(attributeAddress, dataObject);
+        conn.getDlmsMessageListener().setDescription(
+                "SetConfigurationObject AttributeAddress: " + JdlmsObjectToStringUtil.describeAttributes(
+                        attributeAddress));
+        return this.getAccessResultCode(conn, setParameter);
+    }
+
+    private AccessResultCode getAccessResultCode(final DlmsConnectionManager conn, final SetParameter setParameter) {
+        LOGGER.info("Set ConfigurationObject using SetParameter {}", setParameter);
+        try {
+            return conn.getConnection().set(setParameter);
+        } catch (final IOException e) {
+            throw new ConnectionException(e);
+        }
+    }
+
+    abstract DataObject buildSetParameterData(ConfigurationObjectDto configurationToSet,
+            ConfigurationObjectDto configurationOnDevice) throws ProtocolAdapterException;
+
+    BitString getFlags(final ConfigurationObjectDto configurationToSet,
+            final ConfigurationObjectDto configurationOnDevice) throws ProtocolAdapterException {
+        final List<ConfigurationFlagDto> flagsToSet = new ArrayList<>();
+        this.addSettableFlags(configurationToSet, flagsToSet);
+        this.addDeviceFlags(configurationOnDevice, flagsToSet);
+        final byte[] flagBytesToSet = this.toBytes(flagsToSet);
+        return new BitString(flagBytesToSet, NUMBER_OF_FLAG_BITS);
+    }
+
+    private void addSettableFlags(final ConfigurationObjectDto configurationToSet,
+            final List<ConfigurationFlagDto> flagsToSet) {
+        configurationToSet.getConfigurationFlags().getFlags().stream().filter(
+                flagToSet -> !flagToSet.getConfigurationFlagType().isReadOnly()).forEach(flagsToSet::add);
+    }
+
+    // Fill missing flagsToSet with flags from configurationOnDevice
+    private void addDeviceFlags(final ConfigurationObjectDto configurationOnDevice,
+            final List<ConfigurationFlagDto> flagsToSet) {
+        configurationOnDevice.getConfigurationFlags().getFlags().forEach(flagOnDevice -> {
+            if (flagsToSet.stream().noneMatch(
+                    flagToSet -> flagToSet.getConfigurationFlagType() == flagOnDevice.getConfigurationFlagType())) {
+                flagsToSet.add(flagOnDevice);
+            }
+        });
+    }
+
+    private byte[] toBytes(final List<ConfigurationFlagDto> flags) throws ProtocolAdapterException {
+        return this.toBytes(this.toWord(flags));
+    }
+
+    private StringBuilder createEmptyWord() {
+        final StringBuilder word = new StringBuilder();
+        final char[] pad = new char[NUMBER_OF_FLAG_BITS];
+        Arrays.fill(pad, '0');
+        word.append(pad);
+        return word;
+    }
+
+    private String toWord(final List<ConfigurationFlagDto> flags) throws ProtocolAdapterException {
+        final StringBuilder sb = this.createEmptyWord();
+        for (final ConfigurationFlagDto flag : flags) {
```

Comment:
consider changing to stream.filter(f -> f.isEnabled).map().Collectors.join()

Reply:
That would mean wrapping the checked ProtocolAdapterException in a RuntimeException which seems too much overhead to me.

```
        flags.stream().filter(ConfigurationFlagDto::isEnabled).forEach(flag -> {
            final ConfigurationFlagTypeDto flagType = flag.getConfigurationFlagType();
            final Integer bitPosition = this.getBitPosition(flagType).orElseThrow(() -> new ProtocolAdapterException(
                    String.format("ConfigurationFlagTypeDto %s not known for protocol", flagType)));
            sb.setCharAt(bitPosition, '1');
        });
```